### ISSUE TITLE
Apply Linting Only to Files Updated in `git`

### ISSUE THREAD

**Issue by palisadoes** [PalisadoesFoundation/talawa-api#1548]
We need the ability to selectively adjust our linting rules so that they are eventually applied to our entire code base.

1. Linting rules are managed by the `.eslintrc.json` file.
1. They are applied to the entire code base at once.
1. If we change a rule from `warn` to `error` then every file in the code base will be checked and we have frequently seen our tests fail for dozens of files in hundreds of lines of code.

This is not acceptable.

We have been trying to more strictly lint the entire code base manually, but it has been very slow.

## At the user CLI
We need a way to edit `.eslintrc.json` with increasingly stricter linting rules in such a way that the tests are only applied to the changed files with every `git commit` and `git push` command. 

These `git` commands must fail if the linting fails.

## With Every Pull Request

We recently merged a Talawa-API pull request that checks each submitted file for linting compliance. This can be used as an example of how to run the test against individual files.

- https://github.com/PalisadoesFoundation/talawa-admin/pull/1242

You must make sure that this functionality applied for all PRs for this repository whenever the  `.eslintrc.json` file is updated

## Other

1. There may be ways to achieve the CLI goals by updating our existing Git Hooks
1. Your solution should preferably work within coding environments such as VScode
1. Please work with @anwersayeed and @tasneemkoushar on this issue
1. Related issues:
    - https://github.com/PalisadoesFoundation/talawa-admin/issues/1270

## Screenshots

When complete, you should be able to convert all these  `.eslintrc.json` file lines containing `warn` to `error` and the tests will only be applied to the changed files in `git`

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/40879824-664c-4d6f-9906-7372cf80f848)

**Comment by Min2who** [comment]
can i work on it?

**Comment by palisadoes** [comment]
@Nitya-Pasrija Please be aware of this issue

**Comment by NamitBhutani** [comment]
Hi! Can I work on this issue?

**Comment by NamitBhutani** [comment]
I have setup and initialised lint-staged for the repo, in order to convert lint rules from "warn" to "error" , should it be done right now or should I wait for approval? @anwersayeed @tasneemkoushar

**Comment by NamitBhutani** [comment]
@palisadoes I have setup lint-staged (to apply linting to only changed files for every commit). Should I proceed with changing the lint rules from 'warn' to 'error', as I am not sure if this is what was intended to be fixed by this issue.

**Comment by palisadoes** [comment]
Yes, we want:

1. all the warn rules converted to error
2. linting to run correctly, but only on committed files in the local repository

**Comment by palisadoes** [comment]
@NamitBhutani 
The linting is failing for this PR and probably all new PR updates. Please rectify quickly or we will need to revert your PR

1. https://github.com/PalisadoesFoundation/talawa-api/pull/1647
2. https://github.com/PalisadoesFoundation/talawa-api/actions/runs/7442918033/job/20247042509?pr=1647

**Comment by NamitBhutani** [comment]
> @NamitBhutani 
> The linting is failing for this PR and probably all new PR updates. Please rectify quickly or we will need to revert your PR:
> 
> - https://github.com/PalisadoesFoundation/talawa-api/pull/1635
> 
> Context
> 
> 1. https://github.com/PalisadoesFoundation/talawa-api/pull/1647
> 2. https://github.com/PalisadoesFoundation/talawa-api/actions/runs/7442918033/job/20247042509?pr=1647

Extremely sorry! I'll look into it.
Again, sorry for the trouble!

**Comment by NamitBhutani** [comment]
@palisadoes The reason the tests are failing for each PR (including this one) is because we have edited the eslintrc.json, thus when npm run lint:check is ran, it uses the updated eslint rc file with errors instead of warns. We will need to update the github actions or make another eslintrc.json file for github workflows.
![image](https://github.com/PalisadoesFoundation/talawa-api/assets/26519624/42f92089-10fe-4aa9-b56b-3233fce33bfa)

**Comment by palisadoes** [comment]
The GitHub action used errors before. Browse the repository a week ago to see the differences in the files you mention then you'll be better able to see what happened

It definitely used errors before, and ran against each file submitted in the PR as part of a for loop

**Comment by NamitBhutani** [comment]
> The GitHub action used errors before. Browse the repository a week ago to see the differences in the files you mention then you'll be better able to see what happened
> 
> It definitely used errors before, and ran against each file submitted in the PR as part of a for loop

According to my understanding the job "Check for linting, formatting, and type errors"  uses @actions/checkout and then runs the npm run lint:check which is essentially eslint . --max-warnings=1500, i.e it uses the eslintrc.json and as I have edited it to contain errors instead of warnings, that is why the action is erroring.

**Comment by palisadoes** [comment]
How can we get this to work in the same way as proposed in Talawa Admin where we lint the code with increasing severity until we get it to the desired standard? Here is what they are planning.
- https://github.com/PalisadoesFoundation/talawa-admin/pull/1397#issuecomment-1882443722

**Comment by palisadoes** [comment]
For example would it be easier to split this up into multiple issues each handling a different linting rule?

**Comment by NamitBhutani** [comment]
> For example would it be easier to split this up into multiple issues each handling a different linting rule?

I feel if we want to keep the PR workflow consistent, we can rather make a separate eslintrc.json for running on staged files (like I mentioned before but not for the PR workflow) with errors and revert the current eslintrc.json to have "warns" instead.

**Comment by palisadoes** [comment]
1. Unfortunately displaying warnings doesn't lead to action. Errors do. 
2. The only way to independently ensure that the linting is done is for the PR workflow to do it in addition to the git workflow on the local branch

Can we do linting for files committed in the local branch with errors, and do the same for the updated files in the PR? If not, then we'll need to create issues per lint rule and close this one.

Can you implement the incremental approach in both scenarios? It hasn't looked promising so far.

**Comment by NamitBhutani** [comment]
> 1. Unfortunately displaying warnings doesn't lead to action. Errors do. 
> 2. The only way to independently ensure that the linting is done is for the PR workflow to do it in addition to the git workflow on the local branch
> 
> Can we do linting for files committed in the local branch with errors, and do the same for the updated files in the PR? If not, then we'll need to create issues per lint rule and close this one.
> 
> Can you implement the incremental approach in both scenarios? It hasn't looked promising so far.

I will try my best to update the PR workflow too , to accommodate for updated files only instead of all the files!

**Comment by DMills27** [comment]
Consider adding autohooks to the repo after the other work has been completed in this issue, as it will make the repo more manageable in the long run. It is not at all difficult to do and shouldn't take up any time.

**Comment by palisadoes** [comment]
@NamitBhutani This wasn't correctly implemented
- https://github.com/PalisadoesFoundation/talawa-api/issues/1784

**Comment by NamitBhutani** [comment]
> @NamitBhutani This wasn't correctly implemented
> 
>     * [Correctly Lint Files in our PRs #1784](https://github.com/PalisadoesFoundation/talawa-api/issues/1784)
Extremely sorry for the oversight @palisadoes , it was working the last time I checked, but after testing it, the PR workflow had one small issue, the wildcard needs to be after the commit hashes.
   -Before
   `npx eslint $(git diff --name-only --diff-filter=d -- '***.ts' ${{ github.event.before }} ${{ github.sha }})`
  -After
   `npx eslint $(git diff --name-only --diff-filter=d ${{ github.event.before }} ${{ github.sha }}) -- '***.ts' `

Barring this another reason that the lint errors made it past the local pre commit husky checks might be due to contributors making a branch (for a feature) off the main branch and not develop, the main branch does not have the updated stricter linting rules.
Rest assured, after making the changes the Workflow run should catch all the linting errors!

**Comment by palisadoes** [comment]
1. Please open a PR ASAP
2. This seems to have impacted our Family Member PR by @AdityaRaimec22 
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1787

**Comment by palisadoes** [comment]
Is this in the `pull-requests.yml` file?

**Comment by palisadoes** [comment]
1. I just edited it. 
2. Please check if this is correct
    1.  https://github.com/PalisadoesFoundation/talawa-api/blob/develop/.github/workflows/pull-request.yml

**Comment by NamitBhutani** [comment]
> 1. I just edited it.
> 
>     2. Please check if this is correct
>        
>        1. https://github.com/PalisadoesFoundation/talawa-api/blob/develop/.github/workflows/pull-request.yml

Yes! I think this should do it.

**Comment by NamitBhutani** [comment]
> 1. I just edited it.
> 
>     2. Please check if this is correct
>        
>        1. https://github.com/PalisadoesFoundation/talawa-api/blob/develop/.github/workflows/pull-request.yml

Sorry, You need to include 
npx eslint $(git diff --name-only --diff-filter=d 76b087419b69dc1f73c9cabfb04db087dc86ba22 4d0c2db37d09fc7dbf03b7e06afe5648783dc3b0) -- '***.ts'
-- '***.ts' inside the brackets, i.e
npx eslint $(git diff --name-only --diff-filter=d 76b087419b69dc1f73c9cabfb04db087dc86ba22 4d0c2db37d09fc7dbf03b7e06afe5648783dc3b0 -- '***.ts')

**Comment by NamitBhutani** [comment]
I have made the PR @palisadoes

**Comment by palisadoes** [comment]
Why wasn't this discovered when doing the local commits?

**Comment by palisadoes** [comment]
It's merged

**Comment by NamitBhutani** [comment]
> 2\. warned
As I mentioned this *could* be due to them making a branch for their feature of a branch that is not develop and thus does not have the stricter eslint code.
Or the hook was bypassed using `git commit --no-verify`

**Comment by palisadoes** [comment]
1. I've asked @AdityaRaimec22 to join this discussion if he has any further difficulty. 
2. The two of you should be able to work this out together
3. Ask @EshaanAgg or @noman2002 for assistance if you need something merged quickly. I'm going to bed.

**Comment by palisadoes** [comment]
@NamitBhutani 

1. Download the files for this PR and see if there is a linting issue with diffed files, just to be safe.
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1787

**Comment by palisadoes** [comment]
@NamitBhutani Your fix is failing it seems.
- https://github.com/PalisadoesFoundation/talawa-api/actions/runs/7772162538/job/21194271619?pr=1787

**Comment by palisadoes** [comment]
We can't merge Family Members until this is fixed unfortuately

**Comment by AdityaRaimec22** [comment]
> @NamitBhutani Your fix is failing it seems.
> 
> * https://github.com/PalisadoesFoundation/talawa-api/actions/runs/7772162538/job/21194271619?pr=1787
> 
> ![image](https://private-user-images.githubusercontent.com/16875803/302081766-2318ead8-7308-47d1-a426-ca2ac6410a37.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcwMjgyMDUsIm5iZiI6MTcwNzAyNzkwNSwicGF0aCI6Ii8xNjg3NTgwMy8zMDIwODE3NjYtMjMxOGVhZDgtNzMwOC00N2QxLWE0MjYtY2EyYWM2NDEwYTM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA0VDA2MjUwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIwZTJiZjZmNGUwNzA2OWIyODk3NWYzNmZhMmI1ZThhNjljNGY5MmZiYzkzMjc1MGM3ZWEzMTU3OWMyYTUyZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9URv7RBa510GAdpyqbcjm5y87S_QjQnpy0yW96vWGFg)

It is passed check one more time. 

<img width="960" alt="image" src="https://github.com/PalisadoesFoundation/talawa-api/assets/118007662/44d5b129-700b-4160-a69b-075097a58aaf">

**Comment by palisadoes** [comment]
Click on the arrow and you see that it failed, but didn't stop the workflow

**Comment by palisadoes** [comment]
We have a similar PR linting feature in Admin. Check that YAML file for inspiration

**Comment by NamitBhutani** [comment]
> @NamitBhutani Your fix is failing it seems.
> 
>     * https://github.com/PalisadoesFoundation/talawa-api/actions/runs/7772162538/job/21194271619?pr=1787
> 
> 
> ![image](https://private-user-images.githubusercontent.com/16875803/302081766-2318ead8-7308-47d1-a426-ca2ac6410a37.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcwMzM5MzksIm5iZiI6MTcwNzAzMzYzOSwicGF0aCI6Ii8xNjg3NTgwMy8zMDIwODE3NjYtMjMxOGVhZDgtNzMwOC00N2QxLWE0MjYtY2EyYWM2NDEwYTM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA0VDA4MDAzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5ZTI5ZTIzNDFlYjhiNjI5NWY3YTYwNjQzZThmMWJhZDcxYzE3ZGI1N2VjMTIwYmRkOTJiMGQ0MzA2NGRlMjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.IjcUeBggyZ1WMKKFe7UUxQS6oWi_LpW1g2QMBJHzGe0)

I feel that this is something to do with how its fast-forwarding the branch with the develop change we made to fix it, I have one more solution, we can try that too, if you allow.
The solution being using `git diff --name-only ${{ github.sha }}^1` instead of the two commit hashes, as the latest hash is not present in the commit history.

**Comment by NamitBhutani** [comment]
> > @NamitBhutani Your fix is failing it seems.
> > ```
> > * https://github.com/PalisadoesFoundation/talawa-api/actions/runs/7772162538/job/21194271619?pr=1787
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ![image](https://private-user-images.githubusercontent.com/16875803/302081766-2318ead8-7308-47d1-a426-ca2ac6410a37.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcwMzM5MzksIm5iZiI6MTcwNzAzMzYzOSwicGF0aCI6Ii8xNjg3NTgwMy8zMDIwODE3NjYtMjMxOGVhZDgtNzMwOC00N2QxLWE0MjYtY2EyYWM2NDEwYTM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA0VDA4MDAzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5ZTI5ZTIzNDFlYjhiNjI5NWY3YTYwNjQzZThmMWJhZDcxYzE3ZGI1N2VjMTIwYmRkOTJiMGQ0MzA2NGRlMjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.IjcUeBggyZ1WMKKFe7UUxQS6oWi_LpW1g2QMBJHzGe0)
> 
> I feel that this is something to do with how its fast-forwarding the branch with the develop change we made to fix it, I have one more solution, we can try that too, if you allow. The solution being using `git diff --name-only ${{ github.sha }}^1` instead of the two commit hashes, as the latest hash is not present in the commit history.

@EshaanAgg should I make a PR right now if you will be able to merge it?

**Comment by LakshyaSatpal** [comment]
@NamitBhutani Just to help you.
In my opinion, you could try just doing `npm run lint-staged` here, instead of manually checking the changed files. 
![image](https://github.com/PalisadoesFoundation/talawa-api/assets/81241551/52e80133-3772-4533-8e5b-c1c0002461ef)

`npm run lint-staged` works well in pre-commit husky file, so I think it can easily solve this problem.

**Comment by NamitBhutani** [comment]
Lint-staged only works in the pre-commit situation, i.e. when the files are staged ( as the name suggests). As here no files are staged, it won't work.

**Comment by palisadoes** [comment]
Should we then lint on commit too, to prevent committing from being done?

**Comment by LakshyaSatpal** [comment]
> Should we then lint on commit too, to prevent committing from being done?

We already have that in pre-commit file.

**Comment by palisadoes** [comment]
@NamitBhutani Are you sure this is working? We need to merge this PR, that initially discovered the problem
- https://github.com/PalisadoesFoundation/talawa-api/pull/1787

**Comment by NamitBhutani** [comment]
> Please also verify in Talawa Admin. I'd like to know how to standardize the typescript linting workflows

Is it in the main or the develop branch of Talawa admin ?

**Comment by palisadoes** [comment]
Develop

**Comment by NamitBhutani** [comment]
> Develop

They are using pre-made GitHub actions for checking the changed files, while we are using the git diff command in talawa-api.

**Comment by palisadoes** [comment]
1. Now that we know this method works for PRs we should open an issue to do the same in Admin and run a similar testing regime that we just followed to verify it.
2. Are you willing to do that?

**Comment by palisadoes** [comment]
Actually, we have something similar created. You may want to provide some guidance here
- https://github.com/PalisadoesFoundation/talawa-admin/pull/1523
- https://github.com/PalisadoesFoundation/talawa-api/issues/1548

**Comment by NamitBhutani** [comment]
> 1. Now that we know this method works for PRs we should open an issue to do the same in Admin and run a similar testing regime that we just followed to verify it.
> 2. Are you willing to do that?

Yes! Sure

**Comment by NamitBhutani** [comment]
> Actually, we have something similar created. You may want to provide some guidance here
> - https://github.com/PalisadoesFoundation/talawa-admin/pull/1523
> - https://github.com/PalisadoesFoundation/talawa-api/issues/1548

Sure, I will look into it.

**Comment by devsargam** [comment]
@NamitBhutani I am the one who opened the pr
https://github.com/PalisadoesFoundation/talawa-admin/pull/1523

Can you provide some guidance???

**Comment by palisadoes** [comment]
@NamitBhutani Your git linting seems to be failing again. Pleases take a look.
- https://github.com/PalisadoesFoundation/talawa-api/pull/1787

**Comment by NamitBhutani** [comment]
> @NamitBhutani I am the one who opened the pr [PalisadoesFoundation/talawa-admin#1523](https://github.com/PalisadoesFoundation/talawa-admin/pull/1523)
> 
> Can you provide some guidance???

I feel that using git diff is a bad approach as there are a lot of exceptions that rise up, we'll have to change that from the talawa-api too. We need to use the github action already being used in talawa-admin. You can look into the documentation https://github.com/marketplace/actions/changed-files#outputs- and implement it for talawa-admin.

**Comment by NamitBhutani** [comment]
> @NamitBhutani Your git linting seems to be failing again. Pleases take a look.
> 
> * [Added Support for User Family Members #1787](https://github.com/PalisadoesFoundation/talawa-api/pull/1787)

I'll soon make another PR to try fixing it.

**Comment by devsargam** [comment]
@NamitBhutani @palisadoes what about linting on only added files and not modified fields with a seperate eslint config especially for the action???

Example:
If `foo.ts` was modified then we ignore it
but if `bar.ts` was added first time in git we lint it with strict mode????
